flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StateAssignmentOperation.java

```diff
@@ -105,13 +106,70 @@ public boolean assignStates() throws Exception {
 				continue;
 			}
 
-			assignAttemptState(task.getValue(), operatorStates);
+			Set<Integer> executionVertexIndices = new HashSet<>();
+			for (ExecutionVertex executionVertex : task.getValue().getTaskVertices()) {
+				executionVertexIndices.add(executionVertex.getParallelSubtaskIndex());
+			}
+			assignAttemptState(task.getValue(), operatorStates, executionVertexIndices);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Assign states to given execution vertices.
+	 */
+	public boolean assignStates(List<ExecutionVertex> executionVertices) throws Exception {
```

Comment:
There is a lot of duplicated code between this method and the orginal `assignStates()`. I am sure that this is not required and if we rethink this a bit the old method should just be able to call the new one. Duplicating most of the code is not very maintainable. I suggest to rething what we gibe to the state assignent operation in constructor or as parameter to this method and unify it. From a quick look I am very sure this is easily possible.